File: go/perfect-numbers/perfect_numbers.go
package perfect

import (
	"errors"
)

// Classification type
type Classification int

func GetAliquotSum(n int64) int64 {
	var aliquotSum int64
	for i := int64(1); i <= n/2; i++ {
		if n%i == 0 {
			aliquotSum += i
		}
	}
	return aliquotSum
}

const (
	ClassificationAbundant Classification = iota
	ClassificationDeficient
	ClassificationPerfect
)

var ErrOnlyPositive = errors.New("not a positive")

func Classify(n int64) (Classification, error) {
	if n <= 0 {
		return -1, ErrOnlyPositive
	}

	aliquotSum := GetAliquotSum(n)
	switch {
	case aliquotSum > n:
		return ClassificationAbundant, nil
	case aliquotSum < n || aliquotSum == 1:
		return ClassificationDeficient, nil
	default:
		return ClassificationPerfect, nil
	}
}

File: lua/perfect-numbers/perfect-numbers.lua
local function aliquot_sum(n)
  local sum = 0
  for i = 1, math.floor(n / 2), 1 do
    if n % i == 0 then
      sum = sum + i
    end
  end
  return sum
end

local function classify(n)
  local sum = aliquot_sum(n)
  if sum > n then
    return "abundant"
  elseif sum < n or sum == 1 then
    return "deficient"
  else
    return "perfect"
  end
end

return {
  aliquot_sum = aliquot_sum,
  classify = classify
}
